README.md

# ExhaustiveSearch <img src="man/figures/logo.svg" align="right" height = 87/>

<!-- badges: start -->
[![](https://www.r-pkg.org/badges/version/ExhaustiveSearch?color=blue)](https://cran.r-project.org/package=ExhaustiveSearch)
[![R-CMD-check](https://github.com/RudolfJagdhuber/ExhaustiveSearch/workflows/R-CMD-check/badge.svg)](https://github.com/RudolfJagdhuber/ExhaustiveSearch/actions)
<!-- badges: end -->

The aim of this R package is to provide an easy to use, fast and and scalable 
exhaustive search framework. An exhaustive feature selection can require
a very large number of models to be fitted and evaluated. Execution speed and 
memory management are thus crucial factors for such tasks. This package solves 
both issues by using a multi-threaded C++ backend. Memory usage is kept constant
by only storing the best results. This allows evaluations of enormous tasks 
that would normally be seen as unfeasible in standard setups.

## Installation

You can install the release version of the ExhaustiveSearch R package from CRAN:
``` r
install.packages("ExhaustiveSearch")
```

The current development version can be installed from GitHub:
``` r
devtools::install_github("RudolfJagdhuber/ExhaustiveSearch")
```

## Usage

The main function `ExhaustiveSearch()` uses the typical `formula` and `data`
structure, which might be familiar to you from functions like `lm()` or `glm()`. 

As a simple example, an exhaustive linear regression task of the mtcars data set
is analyzed. We evaluate all possible feature combinations to predict the 
response variable 'mpg' (miles per gallon).
``` r
library(ExhaustiveSearch)

data(mtcars)
ES <- ExhaustiveSearch(mpg ~ ., data = mtcars, family = "gaussian")
```
```
Starting the exhaustive evaluation.

 Runtime          |  Completed  |  Status
------------------------------------------
 00d 00h 00m 00s  |  1023/1023  |  100%
------------------------------------------

Evaluation finished successfully.
```
As you can see, the evaluation of 1023 models only took a fraction of a second. 
Printing the result object to the console creates an overview with further 
details to the task.
``` r
print(ES)
```
```
+-------------------------------------------------+
|            Exhaustive Search Results            |
+-------------------------------------------------+
Model family:          gaussian 
Intercept:             TRUE 
Performance measure:   AIC 
Models fitted on:      training set (n = 32)
Models evaluated on:   training set (n = 32)
Models evaluated:      1,023  
Models saved:          1,023 
Total runtime:         00d 00h 00m 00s 
Number of threads:     16 

+-------------------------------------------------+
|                Top Feature Sets                 |
+-------------------------------------------------+
       AIC                  Combination
1 154.1194               wt + qsec + am
2 154.3274          hp + wt + qsec + am
3 154.5631        wt + qsec + am + carb
4 154.9740   disp + hp + wt + qsec + am
5 155.4766                cyl + hp + wt 
```
There are a lot more options to the exhaustive search function, which are 
documented in the help files (see `?ExhaustiveSearch()`). These include:

* logistic regression models for classification
* other performance measures
* limiting the size of combinations (e.g. only up to 5 features)
* defining separate training and testing partitions
* ...


## Performance
The example above is an easy task for this framework and could of course also
be managed by less sophisticated frameworks without greater problems. For larger
tasks however, performance may be an issue. In this section, I will give a crude
estimate of what run-times to expect for different tasks.

The main influencing factors to the run-time are the modeling task, the number
of available threads for parallel execution, the data size and the CPU speed. 

A local benchmark on an AMD Ryzen 7 1700X on a data set of 500 observations
resulted in the following performances:
```
Linear Regression:    45000 models/threadsec
Logistic Regression:   2500 models/threadsec
```
Therefore, on a typical home-PC setup (16 threads), one could expect to be able
to evaluate 2,592,000,000 linear regression models in one hour. For logistic
regression, this number would be 120,000,000 models per hour.

## More about this Package and Exhaustive Searches
In feature and model selection application, exhaustive searches are often 
referred to as *optimal* search strategies, as they test each setup and 
therefore ensure to find the best solution. The main downside of this approach 
is the possibly enormous computational complexity of the task. This package 
provides an easy to use and efficient framework for such problems.

Its main characteristics are:

* Combinations are iteratively generated on the fly,
* Model fitting and evalution is performed multi-threaded in C++,
* Only a fixed amount of models are stored to keep memory usage small.

Currently, ordinary linear regression models similar to `lm()` and logistic 
regression models similar to `glm()` (with parameter family = "binomial") can 
be fitted. All model results of the C++ backend are identical to what would be 
obtained by `glm()` or `lm()`. For that, the logistic regression uses the same 
L-BFGS optimizer as `glm()`.

To assess the quality of a model, the performanceMeasure options 'AIC' (Akaike's
An Information Criterion) and 'MSE' (Mean Squared Error) are implemented.

While this framework is able to handle very large amounts of combinations, an 
exhaustive search of every theoretical combination can still be unfeasible. 
However, a possible way to drastically limit the total number of combinations 
is to define an upper bound for the size of a combination. For example, 
evaluating all combinations of 500 features (`3.3e150`) is obviously 
impossible. But if we only consider combinations of up to 3 features, this 
number reduces to around 21 million, which could easily be evaluated by this 
framework in less than a minute on 16 threads. Setting an upper limit is thus a 
very powerful option to enable high dimensional analyses. It is implemented by 
the 'combsUpTo' parameter.

## Further Development

The official version includes a consistent and ready-to-use framework. 
Nevertheless, a lot more can be done. Further development is managed in this 
GitHub repository. To submit further ideas, please open an issue.


NEWS.md
# ExhaustiveSearch Version 1
## Subversion 1.0
### Patch 1.0.0

* First release version available on CRAN

### Patch 1.0.1

* Fixed missing definition of size_t, which led to errors on solaris.


R/print.R


formatSecTime = function(sec) {
  days = sec / 60 / 60 / 24
  hour = (sec / 60 / 60) %% 24
  min = (sec / 60) %% 60
  osec = sec %% 60
  paste(paste0(sprintf("%02d", floor(c(days, hour, min, osec))),
               c("d ", "h ", "m ", "s")), collapse = "")
}

#' Print ExhaustiveSearch
#'
#' Prints a compact summary of the results of an ExhaustiveSearch object.
#'
#' @param x Object of class 'ExhaustiveSearch'.
#' @param ... Further arguments passed to or from other methods.
#'
#' @return No return value. The function is only called to print results to the
#'   console.
#'
#' @author Rudolf Jagdhuber
#'
#' @seealso [ExhaustiveSearch()]
#'
#' @importFrom utils capture.output
#' @export
print.ExhaustiveSearch = function(x, ...) {

  evalOn = ifelse(x$setup$nTest == 0,
    paste0("training set (n = ", format(x$setup$nTrain, big.mark = ","), ")\n"),
    paste0("test set (n = ", format(x$setup$nTest, big.mark = ","), ")\n"))

  cat("\n+-------------------------------------------------+")
  cat("\n|            Exhaustive Search Results            |")
  cat("\n+-------------------------------------------------+\n")
  cat("Model family:         ", x$setup$family, "\n")
  cat("Intercept:            ", x$setup$intercept, "\n")
  cat("Performance measure:  ", x$setup$performanceMeasure, "\n")
  cat("Models fitted on:     ", " training set (n = ", x$setup$nTrain, ")\n",
    sep = "")
  cat("Models evaluated on:  ", evalOn)
  cat("Models evaluated:     ", format(x$nModels, big.mark = ","),
    ifelse(x$evaluatedModels != x$nModels, " (Incomplete!)", ""), "\n")
  cat("Models saved:         ", format(x$setup$nResults, big.mark = ","), "\n")
  cat("Total runtime:        ", formatSecTime(x$runtimeSec), "\n")
  cat("Number of threads:    ", x$batchInfo$nBatches, "\n")
  cat("\n+-------------------------------------------------+")
  cat("\n|                Top Feature Sets                 |")
  cat("\n+-------------------------------------------------+\n")
  cat(paste(capture.output(resultTable(x, 5, "  ")), collapse = "\n"), "\n\n")
}


R/resultFunctions.R

#' Overview of the top exhaustive search results
#'
#' Extract the top `n` results of an exhaustive search and present them as a
#' `data.frame` object.
#'
#' @details
#' The result of an exhaustive search is given by an object of class
#' ExhaustiveSearch, which is a list of encoded feature combinations and
#' performance values. This function decodes the feature combinations and
#' presents them in a `data.frame` together with the respective performance
#' values
#'
#' @param ESResult a result object from an exhaustive search.
#' @param n number of results to be returned. The default (`Inf`) returns every
#'   result available in `ESResult`.
#' @param insertStart used for additional spacing when printing. The value of
#'   `insertStart` gets printed in front of every feature combination to
#'   increase the space to the printed performance measure.
#'
#' @return A `data.frame` with two columns. The first one shows the performance
#'   values and the second shows the decoded feature set collapsed with plus
#'   signs.
#'
#' @examples
#' ## Exhaustive search on the mtcars data
#' data(mtcars)
#' ES <- ExhaustiveSearch(mpg ~ ., data = mtcars, family = "gaussian")
#'
#' ## Summary data.frame of the top 5 models
#' resultTable(ES, 5)
#'
#' ## Return a data.frame of all stored models
#' res <- resultTable(ES)
#' str(res)
#'
#' ## Add custom characters for printing
#' resultTable(ES, 1, "  <->  ")
#'
#' @author Rudolf Jagdhuber
#'
#' @seealso [ExhaustiveSearch()]
#'
#' @export
resultTable = function(ESResult, n = Inf, insertStart = "") {
  n = min(n, length(ESResult$ranking$performance))
  ret = data.frame(ESResult$ranking$performance[seq_len(n)],
    sapply(ESResult$ranking$featureIDs[seq_len(n)],
      function(ids) paste0(insertStart,
        paste(ESResult$featureNames[ids], collapse = " + "))))
  colnames(ret) = c(ESResult$setup$performanceMeasure, "Combination")
  return(ret)
}


#' Extract the feature sets from an ExhaustiveSearch object
#'
#' A simple function to get a vector of feature names for one or more elements
#' of an ExhaustiveSearch object.
#'
#' @param ESResult a result object from an exhaustive search.
#' @param ranks a numeric value or vector defining which elements should be
#'   returned.
#'
#' @return If `ranks` is a single value, a vector of feature names is returned.
#'   If an intercept is included, the first element of this vector is "1". If
#'   `ranks` includes multiple values, a list of such vectors is returned.
#'
#' @examples
#' ## Exhaustive search on the mtcars data
#' data(mtcars)
#' ES <- ExhaustiveSearch(mpg ~ ., data = mtcars, family = "gaussian")
#'
#' ## Get the feature combinations of the top 3 models
#' getFeatures(ES, 1:3)
#'
#' ## Get the feature combination of the 531th best model
#' getFeatures(ES, 531)
#'
#' @author Rudolf Jagdhuber
#'
#' @seealso [ExhaustiveSearch()]
#'
#' @export
getFeatures = function(ESResult, ranks) {
  sapply(ranks, function(x) c(ifelse(ESResult$setup$intercept, "1", NULL),
    ESResult$featureNames[ESResult$ranking$featureIDs[[x]]]))
}


R/ExhaustiveSearch.R


#' Exhaustive feature selection
#'
#' Performs an exhaustive feature selection. `ExhaustiveSearch()` is a fast and
#' scalable implementation of an exhaustive feature selection framework. It is
#' particularly suited for huge tasks, which would typically not be possible due
#' to memory limitations. The current version allows to compute linear and
#' logistic regression models and compare them with respect to AIC or MSE.
#'
#' @details
#' An exhaustive search evaluates all setups of a combinatorial task. In feature
#' and model selection application, exhaustive searches are often referred to as
#' *optimal* search strategies, as they test each setup and therefore ensure to
#' find the best solution. The main downside of this approach is the possibly
#' enormous computational complexity of the task. `ExhaustiveSearch()` provides
#' an easy to use and efficient framework for these tasks.
#'
#' Its main characteristics are:
#' - Combinations are iteratively generated on the fly,
#' - Model fitting and evalution is performed multi-threaded in C++,
#' - Only a fixed amount of models are stored to keep memory usage small.
#'
#' Therefore, the framework of this package is able to evaluate huge tasks of
#' billions of models, while only being limited by run-time.
#'
#' Currently, ordinary linear regression models similar to [lm()] and logistic
#' regression models similar to [glm()] (with parameter `family = "binomial"`)
#' can be fitted. The model type is specified via the `family` parameter. All
#' model results of the C++ backend are identical to what would be obtained by
#' [glm()] or [lm()]. For that, the logistic regression also uses the same
#' \href{https://en.wikipedia.org/wiki/Limited-memory_BFGS}{L-BFGS} optimizer
#' as [glm()].
#'
#' To assess the quality of a model, the `performanceMeasure` options 'AIC'
#' (Akaike's An Information Criterion) and 'MSE' (Mean Squared Error) are
#' implemented. Note that the AIC can only be computed on the training data,
#' while it is recommended for the MSE to be computed on independent test data.
#' If `performanceMeasure` is not set, it will be decided according to the
#' definition of a test data set.
#'
#' While this framework is able to handle very large amounts of combinations, an
#' exhaustive search of every theoretical combination can still be unfeasible.
#' However, a possible way to drastically limit the total number of combinations
#' is to define an upper bound for the size of a combination. For example,
#' evaluating all combinations of 500 features (3.3e150) is obviously
#' impossible. But if we only consider combinations of up to 3 features, this
#' number reduces to around 21 million, which could easily be evaluated by this
#' framework in less than a minute (16 threads). Setting an upper limit is thus
#' a very powerful option to enable high dimensional analyses. It is implemented
#' by the parameter `combsUpTo`.
#'
#' A core element of why this framework does not require more memory if tasks
#' get larger is that at any point the best models are stored in a list of
#' fixed size. Therefore, sub-optimal models are not saved and do not take space
#' and time to be handled. The parameter defining the size of the models, which
#' are actively stored is `nResults`. Large values here can impair performance
#' or even cause errors, if the system memory runs out and should always be set
#' with care. The function will however warn you beforehand if you set a very
#' large value here.
#'
#' The parameter `testSetIDs` can be used to split the data into a training and
#' testing partition. If it is not set, all models will be trained and tested on
#' the full data set. If it is set, the data will be split beforehand into
#' `data[testSetIDs,]` and `data[-testSetIDs,]`.
#'
#'
#'
#' The development version of this package can be found at
#' \url{https://github.com/RudolfJagdhuber/ExhaustiveSearch}. Issues or requests
#' are handled on this page.
#'
#' @param formula An object of class [formula] (or one that can be coerced to
#'   that class): a symbolic description of the feature and response structure.
#'   All combinations of features on the right hand side of the formula are
#'   evaluated.
#' @param data A [data.frame] (or object coercible by [as.data.frame()] to a
#'   [data.frame]) containing the variables in the model.
#' @param family A [character] string naming the family function similar to the
#'   parameter in [glm()]. Currently options are 'gaussian' or 'binomial'. If
#'   not specified, the function tries to guess it from the response variable.
#' @param performanceMeasure A [character] string naming the performance measure
#'   to compare models by. Currently available options are 'AIC' (Akaike's An
#'   Information Criterion) or 'MSE' (Mean Squared Error).
#' @param combsUpTo An integer of length 1 to set an upper limit to the number
#'   of features in a combination. This can be useful to drastically reduce the
#'   total number of combinations to a feasible size.
#' @param nResults An integer of length 1 to define the size of the final
#'   ranking list. The default (5000) provides a good trade-off of memory usage
#'   and result size. Set this value to `Inf` to store all models.
#' @param nThreads Number of threads to use. The default is to detect the
#'   available number of threads automatically.
#' @param testSetIDs A vector of row indices of data, which define the test set
#'   partition. If this parameter is `NULL` (default), models are trained and
#'   evaluated on the full data set. If it is set, models are trained on
#'   `data[-testSetIDs,]` and tested on `data[testSetIDs,]`.
#' @param errorVal A numeric value defining what performance result is returned
#'   if the model could not be fitted. The default (-1) makes those models
#'   appear at the top of the result ranking.
#' @param quietly [logical]. If set to `TRUE` (default), status and runtime
#'   updates are printed to the console.
#' @param checkLarge [logical]. Very large calls get stopped by a safety net.
#'   This parameter can be used to execute these calls anyway.
#'
#' @return Object of class `ExhaustiveSearch` with elements
#'   \item{nModels}{The total number of evaluated models.}
#'   \item{runtimeSec}{The total runtime of the exhaustive search in seconds.}
#'   \item{ranking}{A list of the performance values and the featureIDs. The
#'     i-th element of both correspond. The featureIDs refer to the elements of
#'     `featureNames`. Formatted results of these rankings can e.g. be obtained
#'     with [getFeatures()], or [resultTable()].}
#'   \item{featureNames&#160;&#160;}{The feature names in the given data.
#'     `featureIDs` in the ranking element refer to this vector.}
#'   \item{batchInfo}{A list of information on the batches, into which the
#'     total task has been partitioned. List elements are the number of batches,
#'     the number of elements per batch, and the combination boundaries that
#'     define the batches.}
#'   \item{setup}{A list of input parameters from the function call.}
#'
#' @examples
#' ## Linear Regression on mtcars data
#' data(mtcars)
#'
#' ##  Exhaustive search of 1023 models compared by AIC
#' ES <- ExhaustiveSearch(mpg ~ ., data = mtcars, family = "gaussian",
#'   performanceMeasure = "AIC")
#' print(ES)
#'
#' ## Same setup, but compared by MSE on a test set partition
#' testIDs <- sample(nrow(mtcars), round(1/3 * nrow(mtcars)))
#' ES2 <- ExhaustiveSearch(mpg ~ ., data = mtcars, family = "gaussian",
#'   performanceMeasure = "MSE", testSetIDs = testIDs)
#' print(ES2)
#'
#'
#' \dontrun{
#' ## Logistic Regression on Ionosphere Data
#' data("Ionosphere", package = "mlbench")
#'
#' ## Only combinations of up to 3 features! -> 5488 models instead of 4 billion
#' ES3 <- ExhaustiveSearch((Class == "good") ~ ., data = Ionosphere[,-c(1, 2)],
#'   family = "binomial", combsUpTo = 3)
#' print(ES3)
#' }
#'
#' @author Rudolf Jagdhuber
#'
#' @seealso [resultTable()], [getFeatures()]
#'
#' @importFrom Rcpp evalCpp
#' @import stats
#' @export
ExhaustiveSearch = function(formula, data, family = NULL,
  performanceMeasure = NULL, combsUpTo = NULL, nResults = 5000, nThreads = NULL,
  testSetIDs = NULL, errorVal = -1, quietly = FALSE, checkLarge = TRUE) {

  formula = formula(formula)
  if (!inherits(formula, "formula")) stop("\nInvalid formula.")

  data = as.data.frame(data)
  if (!inherits(data, "data.frame")) stop("\nInvalid data object.")

  ## Extract the design matrix with dims, response vector and feature names
  X = stats::model.matrix(formula, data)

  ## Does the setup include an intercept
  intercept = attr(terms(formula, data = data), "intercept") == 1

  ## Features combinations are stored as 1, 2, 3,... in C++, 0 is reserved for
  ## the intercept, which is the first columnof the DataSet (Data[,0]). To keep
  ## the C++ subsetting via Column[,index] consistent for setups with and
  ## without intercepts, an unused dummy column needs to be added at setups
  ## without intercept. This makes Data[,i] always refer to the same feature i.
  if (!intercept) X = cbind("NotUsed" = 0, X)

  feats = colnames(X)[-1]
  y = as.numeric(model.response(model.frame(formula, data)))

  ## Split into training and testing partitions
  if (length(testSetIDs) == 0) {
    XTest = X[NULL, ]
    yTest = y[NULL]
  } else if (!is.numeric(testSetIDs) | any(testSetIDs > nrow(X)) |
      length(testSetIDs) >= nrow(X)) {
    stop(paste0(
      "\nThe given testSetIDs need to be numeric and cannot exceed the data\n",
      "dimension.\n\n"))
  } else {
    XTest = X[testSetIDs, ]
    yTest = y[testSetIDs]
    X = X[-testSetIDs,]
    y = y[-testSetIDs]
  }

  ## Check if the family parameter was set correctly
  if (is.null(family)) {
    family = ifelse(length(unique(y)) == 2, "binomial", "gaussian")
    warning(paste0("\n\n",
      "'family' not specified! From the given response data, I assume\n'",
      family, "' and continue.\n\n"))
  } else if (family == "gaussian") {
    if (length(unique(y)) == 2) warning(paste0("\n\n",
      "'family' was set to 'gaussian', but the response appears to be\n",
      "binary, are you sure this is intended?\n\n"))
  } else if (family == "gaussian") {
    if (length(unique(y)) != 2) stop(paste0("\n",
      "'family' was set to 'binomial', but the response is not binary.\n\n"))
  }

  ## Check the performanceMeasure parameter and the validity of each case
  if (is.null(performanceMeasure)) {
    if (nrow(XTest) == 0) performanceMeasure = "AIC"
    else performanceMeasure = "MSE"
  } else if (performanceMeasure == "AIC") {
    if (nrow(XTest) != 0) stop(paste0("\n",
      "A TestSet was defined, but 'performanceMeasure' is set to 'AIC'.\n\n"))
  } else if (performanceMeasure == "MSE") {
    if (nrow(XTest) == 0) warning(paste0("\n\n",
      "No TestSet was defined and performanceMeasure set to 'MSE'.\n",
      "Comparing MSE values on training data will always prefer the higher\n",
      "dimensional model in nested setups and is thus not recommended for\n",
      "feature selection tasks.\n\n"))
  } else stop(paste0("\n",
    "Unsupported performanceMeasure! Please check the help file for a list\n",
    "of all available options.\n\n"))

  ## Check combUpTo parameter
  if (is.null(combsUpTo)) combsUpTo = length(feats)
  if (!is.numeric(combsUpTo) | length(combsUpTo) != 1 | any(combsUpTo <= 0))
    stop("\ncombsUpTo needs to be a single numeric value > 0\n\n")
  if (combsUpTo == Inf) combsUpTo = length(feats)


  ## Safety-check if the user requests a huge task
  nCombs = sum(choose(length(feats), seq(combsUpTo)))
  if (nCombs > 1e8 & checkLarge) stop(paste0(
    "\nThe requested task needs to evaluate a huge number of combinations:\n\n",
    "  -> ", format(nCombs, big.mark = ",", scientific = FALSE)," models.\n\n",
    "Consider reducing the total number of combinations with 'combsUpTo'.\n\n",
    "To continue with this setup, set the parameter 'checkLarge = FALSE'.\n\n"))

  ## Check nResults parameter
  if (is.null(nResults)) nResults = nCombs
  if (!is.numeric(nResults) | length(nResults) != 1 | any(nResults <= 0))
    stop("\nnResults needs to be a single numeric value > 0")
  if (nResults > nCombs) nResults = nCombs

  # Check nThreads parameter, if not set detect later in C++
  if (is.null(nThreads)) nThreads = 0
  else if (!is.numeric(nThreads) | length(nThreads) != 1 | nThreads %% 1 != 0)
    stop("\nnThreads needs to be a single integer value\n\n")

  ## Check errorVal parameter
  if (is.null(errorVal)) stop("\nerrorVal parameter cannot be NULL\n\n")
  if (!is.numeric(errorVal) | length(errorVal) != 1)
    stop("\nerrorVal needs to be a single numeric value\n\n")

  if (!quietly) cat("\nStarting the exhaustive evaluation.\n\n")

  ## The main C++ function call
  cppOutput = ExhaustiveSearchCpp(
    XInput = X,
    yInput = y,
    XTestSet = XTest,
    yTestSet = yTest,
    family = family,
    performanceMeasure = performanceMeasure,
    intercept = intercept,
    combsUpTo = combsUpTo,
    nResults = nResults,
    nThreads = nThreads,
    errorVal = errorVal,
    quietly = quietly)

  if (length(cppOutput) == 0) stop("\n\nAn internal error occured.\n\n")

  ## Preparation of the return value
  result = list()
  class(result) = "ExhaustiveSearch"

  result$nModels = cppOutput[[4]]
  result$runtimeSec = cppOutput[[1]]
  result$ranking = list(
    performance = cppOutput[[2]],
    featureIDs = cppOutput[[3]])
  result$featureNames = feats
  result$batchInfo = list(nBatches = cppOutput[[5]],
    batchSizes = cppOutput[[6]], batchLimits = cppOutput[[7]])
  result$setup = list(call = match.call(), family = family,
    performanceMeasure = performanceMeasure, intercept = intercept,
    combsUpTo = combsUpTo, nResults = nResults, nThreads = nThreads,
    testSetIDs = testSetIDs, nTrain = nrow(X), nTest = nrow(XTest))

  if (!quietly) {
    if (cppOutput[[4]] == nCombs) cat("\nEvaluation finished successfully.\n\n")
    else warning("\n\nEvaluation Incomplete! Not all models were evaluated!\n")
  }

  return(result)
}
